Add filter and scroll for EventSource selector

rottencandy opened the pull request:
**Fixes**: 
https://issues.redhat.com/browse/ODC-4244
<!-- For e.g Fixes: https://issues.redhat.com/browse/ODC-XXX -->

**Analysis / Root cause**: 
As a user, i would like the ability to filter eventSources by name and scroll through list if goes more than two rows.
<!-- Briefly describe analysis of US/Task or root cause of Defect -->

**Solution Description**: 
This PR adds the following changes:
- Add a fuzzy text filter to event sources list
- Add an item counter on the top right
- Disable default selection on page load
<!-- Describe your code changes in detail and explain the solution -->

**Screen shots / Gifs for design review**: 
![ev1](https://user-images.githubusercontent.com/20013884/87025180-aeaede00-c1f7-11ea-8521-c58774a5f8b1.png)
![ev2](https://user-images.githubusercontent.com/20013884/87025226-bec6bd80-c1f7-11ea-9c24-5fd9970a89b1.png)
![ev3](https://user-images.githubusercontent.com/20013884/87025239-c1c1ae00-c1f7-11ea-9602-43cf9ae29ab2.png)

cc: @openshift/team-devconsole-ux @invincibleJai 

<!-- If change affects UI in any way, tag @openshift/team-devconsole-ux and add screenshots/gifs  -->

**Browser conformance**: 
<!-- To mark tested browsers, use [x] -->
- [ ] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge

invincibleJai commented:
should we have `clear all filters` as well in empty state if non matches

![image](https://user-images.githubusercontent.com/5129024/87037373-ba0b0500-c209-11ea-8301-cbdf7636a783.png)

rottencandy commented:
> should we have `clear all filters` as well in empty state if non matches

Initially did not have it as I wasn't using a state for the input field, updated and added a clear button.

sahil143 commented on frontend/packages/knative-plugin/src/components/add/event-sources/EventSourcesSelector.tsx:
```diff
@@ -25,6 +37,8 @@ const EventSourcesSelector: React.FC<EventSourcesSelectorProps> = ({ eventSource
     setStatus,
   } = useFormikContext<FormikValues>();
+  const [filteredSources, setFilteredSources] = React.useState(eventSourceList);
+  const [filterText, setFilterText] = React.useState('');
```
Why do we need this state for `filterText`? State for `filteredSources` should be enough.

rottencandy replied:
State for `filterText` is used to clear the input field when "clear all filters" is clicked from the empty state.

sahil143 commented on frontend/packages/knative-plugin/src/components/add/event-sources/EventSourcesSelector.tsx:
```diff
@@ -53,18 +67,67 @@ const EventSourcesSelector: React.FC<EventSourcesSelectorProps> = ({ eventSource
+  ]);
+
+  const handleFilterChange = (text: string) => {
+    setFilterText(text);
```
remove this.

rottencandy replied:
Moved the filter to `ItemSelectorField`.

sahil143 commented on frontend/packages/knative-plugin/src/components/add/event-sources/EventSourcesSelector.scss:
```diff
@@ -0,0 +1,17 @@
+  align-items: center;
+
+  &__bar {
+    max-width: 20em;
```
Use pf variables

rottencandy replied:
Moved scss to `ItemSelectorField.scss`

sahil143 commented on frontend/packages/knative-plugin/src/components/add/event-sources/EventSourcesSelector.tsx:
```diff
@@ -53,18 +67,67 @@ const EventSourcesSelector: React.FC<EventSourcesSelectorProps> = ({ eventSource
+        placeholder="Filter by type..."
+      />
+      <span className="odc-event-source-filter__items-count">
+        {_.keys(eventSourceList).length} items
```
```suggestion
        {_.keys(filteredSources).length} items
```

rottencandy replied:
Updated.

sahil143 commented on frontend/packages/knative-plugin/src/components/add/event-sources/EventSourcesSelector.tsx:
```diff
@@ -53,18 +67,67 @@ const EventSourcesSelector: React.FC<EventSourcesSelectorProps> = ({ eventSource
+
+  const textFilter = (
+    <div className="odc-event-source-filter">
+      <TextInput
```
Use `InputField` component here. It uses TextInput internally.

sahil143 commented on frontend/packages/knative-plugin/src/components/add/event-sources/EventSourcesSelector.tsx:
```diff
@@ -53,18 +67,67 @@ const EventSourcesSelector: React.FC<EventSourcesSelectorProps> = ({ eventSource
+    </div>
+  );
+
+  const itemSelectorField = _.keys(filteredSources).length ? (
```
Be more explicit
```suggestion
  const itemSelectorField = _.keys(filteredSources).length > 0 ? (
```

rottencandy replied:
Updated.

sahil143 commented:
> should we have `clear all filters` as well in empty state if non matches
> 
> ![image](https://user-images.githubusercontent.com/5129024/87037373-ba0b0500-c209-11ea-8301-cbdf7636a783.png)

@invincibleJai Do we really need the `Clear All Filters` button there because there is only one input field and the user will be already focused on that input field when the empty state is shown. WDYT?
cc @christianvogt

invincibleJai commented:
> > should we have `clear all filters` as well in empty state if non matches
> > ![image](https://user-images.githubusercontent.com/5129024/87037373-ba0b0500-c209-11ea-8301-cbdf7636a783.png)
> 
> @invincibleJai Do we really need the `Clear All Filters` button there because there is only one input field and the user will be already focused on that input field when the empty state is shown. WDYT?
> cc @christianvogt

yeah you are right it's screenshot from catalog and added as it's there in AC for epic as well. If we see from consistency POV then make sense sense to show and if from usability POV as user will be in search field itself and we just have one filter so we can avoid

We need check with UXD or @christianvogt can add input

rachael-phillips commented:
Hey @invincibleJai and @rottencandy can we take a step back and include UX in this process before we move forward with the pr? I am happy to set up a time so we can talk.

rachael-phillips commented:
Is there a reason why we are we introducing the filter now when there are a small number of event sources? We are going to be creating a view that scales to a larger number of event sources when it would be more appropriate to have a filter. If this is planning for the future then I understand, trying to get some background here- could you explain the reasoning?

christianvogt commented on frontend/packages/knative-plugin/src/components/add/event-sources/EventSourcesSelector.scss:
```diff
@@ -0,0 +1,17 @@
+}
+
+.odc-event-source-selector {
+  height: 260px;
```
This adds too much white beneath the selector when there is a single row.
Perhaps max-height is better suited here.

rottencandy replied:
Updated.

christianvogt commented on frontend/packages/knative-plugin/src/components/add/event-sources/EventSourcesSelector.tsx:
```diff
@@ -53,18 +67,67 @@ const EventSourcesSelector: React.FC<EventSourcesSelectorProps> = ({ eventSource
+    debounceFilterText(text);
+  };
+
+  const textFilter = (
```
Would be better to use this filter inside `ItemSelectorField` so that it is reusable if other forms want the selector with a filter.

rottencandy replied:
Moved filter to `ItemSelectorField`, can now be enabled through `showFilter` and `showCount` props.

rottencandy commented on frontend/packages/knative-plugin/src/components/add/event-sources/EventSourcesSelector.tsx:
```diff
@@ -53,18 +67,67 @@ const EventSourcesSelector: React.FC<EventSourcesSelectorProps> = ({ eventSource
+        No Results Match the Filter Criteria
+      </Title>
+      <EmptyStateBody>
+        No EventSource types are being shown due to the filters being applied.
```
Updated.

rottencandy commented on frontend/packages/knative-plugin/src/components/add/event-sources/EventSourcesSelector.tsx:
```diff
@@ -53,18 +67,67 @@ const EventSourcesSelector: React.FC<EventSourcesSelectorProps> = ({ eventSource
+      </EmptyStateBody>
+      <EmptyStatePrimary>
+        <Button variant="link" onClick={() => handleFilterChange('')}>
+          Clear all filters
```
Changed.

rottencandy commented:
Updated screenshots:
![ev1](https://user-images.githubusercontent.com/20013884/87153250-ce680400-c2d4-11ea-9ad8-066da2d931b9.png)
![ev2](https://user-images.githubusercontent.com/20013884/87153256-d1fb8b00-c2d4-11ea-8e27-ddd18bd997e7.png)
![temp](https://user-images.githubusercontent.com/20013884/87153264-d45de500-c2d4-11ea-866e-87a95ba88618.png)

rottencandy commented on frontend/packages/console-shared/src/components/formik-fields/item-selector-field/ItemSelectorField.tsx:
```diff
@@ -94,20 +134,58 @@ const ItemSelectorField: React.FC<ItemSelectorFieldProps> = ({
+              />
+              {showCount && (
+                <span className="odc-item-selector-filter__count">
+                  {itemCount === 1 ? '1 item' : `${itemCount} items`}
```
Updated

rottencandy commented on frontend/packages/console-shared/src/components/formik-fields/item-selector-field/ItemSelectorField.tsx:
```diff
@@ -94,20 +134,58 @@ const ItemSelectorField: React.FC<ItemSelectorFieldProps> = ({
+                className="odc-item-selector-filter__input"
+                onChange={handleFilterChange}
+                value={filterText}
+                placeholder={'Filter by type...'}
```
Changed.

rottencandy commented on frontend/packages/console-shared/src/components/formik-fields/item-selector-field/ItemSelectorField.tsx:
```diff
@@ -36,10 +48,15 @@ const ItemSelectorField: React.FC<ItemSelectorFieldProps> = ({
   autoSelect,
+  showIfSingle,
+  showFilter,
+  showCount,
```
Changed

sahil143 commented:
@christianvogt @invincibleJai Does this look alright? Create button is enabled because it satisfies all the validation, but also has an empty state
<img width="1491" alt="Screenshot 2020-07-15 at 9 51 58 PM" src="https://user-images.githubusercontent.com/9278015/87570085-b08c0c00-c6e5-11ea-9ddc-57b418870ace.png">

christianvogt commented:
I thought UX wanted the select item to _always_ be present

beaumorley commented:
@invincibleJai @christianvogt @sahil143 

Looking at the screen capture in the last comment, is the proposal to show fields below the Empty State message? Is the logic to show APIServerSource fields because it is first?

It looks broken to me. I would not want to submit this form. I may be missing something but I would think the form fields would be hidden and the submit would be canceled. If there is justification for this, I just did not see it in the comments but may have missed it ;).

cc @rachael-phillips

invincibleJai commented:
> @invincibleJai @christianvogt @sahil143
> 
> Looking at the screen capture in the last comment, is the proposal to show fields below the Empty State message? Is the logic to show APIServerSource fields because it is first?

@beaumorley  no it's not as `APIServerSource` is first, assume user selected `PingSource` and based on selection form is shown now user goes and filter `zyx` so in selector it'll be empty still below form will be there. So for this scenario what do you suggest??

christianvogt commented:
I think the only time you will see the "no matches" message is when there is no type selected and the filter has no matches.

So long as there is one type selected, that item will _never_ be filtered out and therefore showing the form below still makes sense.

sahil143 commented:
> I think the only time you will see the "no matches" message is when there is no type selected and the filter has no matches.
> 
> So long as there is one type selected, that item will _never_ be filtered out and therefore showing the form below still makes sense.

@rachael-phillips FYI.

rottencandy commented:
Updated, selected tile will always be shown:
![temp](https://user-images.githubusercontent.com/20013884/87641585-f2f52d80-c765-11ea-9d86-5f591e343d5a.png)

rottencandy commented:
/retest

rachael-phillips commented:
I gave this a review, and it looks good to me.

rottencandy commented:
/retest

christianvogt commented:
@rottencandy when i try this, there is no default selected type. I'm not sure if that was supposed to change or not. However once I start typing in a filter. ApiServerSource randomly gets selected...

christianvogt commented:
/lgtm cancel

rottencandy commented:
> @rottencandy when i try this, there is no default selected type. I'm not sure if that was supposed to change or not. However once I start typing in a filter. ApiServerSource randomly gets selected...

I think I found the bug. Previously, if there was a single item in the list, it would get automatically selected, but it gets picked from the original unfiltered list. I've fixed it and tried to maintain that auto selection behaviour. PTAL.
![demo](https://user-images.githubusercontent.com/20013884/87750655-ab2fde00-c819-11ea-9408-7ea4207b1836.gif)

@christianvogt I thought it was decided to remove the default type selection, or I might have mistaken. I can remove that change if it is not required.

invincibleJai commented:
@rottencandy Saud overall looks good now after the fix :)

With just one source we show item selector but `type` is missing from heading and alert touches the filter

![image](https://user-images.githubusercontent.com/5129024/87778915-8d7a6d00-c849-11ea-8318-80a8240075f8.png)

rottencandy commented:
> @rottencandy Saud overall looks good now after the fix :)
> 
> With just one source we show item selector but `type` is missing from heading and alert touches the filter
Fixed the title, It renders properly now.
> and in console can see one error complaining about `Text input: Text input requires either an id or aria-label to be specified`
Added the `aria-label` to the Text input.

I also rebased the branch with `master`.

invincibleJai commented:
/lgtm

verified locally, seems to work as expected

rottencandy commented:
/retest

rottencandy commented:
/retest

christianvogt commented:
I do not thing we should be automatically selecting a type for the user if there are multiple and they happen to enter a filter value that brings the list down to 1. It's fine if the list began with only a single option.

cc @beaumorley

invincibleJai commented:
> I do not thing we should be automatically selecting a type for the user if there are multiple and they happen to enter a filter value that brings the list down to 1. It's fine if the list began with only a single option.
> 
> cc @beaumorley

In this PR it's not selecting a automatically for the user in case of multiple or if they happen to to enter a filter value that brings the list down to 1

It's preselecting if the list began with only a single option.

rottencandy commented:
Thanks for the clarification. Updated to not select card if a single value is filtered out.
![demo](https://user-images.githubusercontent.com/20013884/87908475-0fee7100-ca84-11ea-9fa4-ec0f44bacbef.gif)

rottencandy commented:
/retest

rottencandy commented:
/retest

invincibleJai commented:
/lgtm

Verified locally, works as expected

rottencandy commented:
/assign @christianvogt

christianvogt commented:
/approve

rachael-phillips commented:
LGTM
